src/site/content/en/blog/racer/index.md

```diff
@@ -0,0 +1,289 @@
+
+```js
+_car.style.webkitTransform = 'translate3d('+_position.x+'px, '+_position.y+'px, 0px)rotate('+_rotation+'deg)';
+```
+
+On Chrome for Android, the best performance was achieved by calculating the matrix values and applying a matrix transform:
+
+```js
+var rad = _rotation.rotation * (Math.PI * 2 / 360);
+var cos = Math.cos(rad);
+var sin = Math.sin(rad);
+var a = parseFloat(cos).toFixed(8);
+var b = parseFloat(sin).toFixed(8);
+var c = parseFloat(-sin).toFixed(8);
+var d = a;
+_car.style.webkitTransform = 'matrix('+a+', '+b+', '+c+', '+d+', '+_position.x+', '+_position.y+')';
+```
+
+## Keeping the Devices Synced
+
+The most important (and difficult) part of development was to make sure the game synced across devices. We thought users could be forgiving if a car occasionally skipped a few frames because of a slow connection but it wouldn’t be much fun if your car is jumping around, appearing on multiple screens at once.  Resolving this required a ton of trial and error, but we eventually settled on a few tricks which made it work.
+
+### Calculating Latency
+
+The starting point for syncing devices is knowing how long it takes for messages to be received from the Compute Engine relay. The tricky part is that the clocks on each device will never be completely in sync. To get around this, we needed to find the difference in time between the device and the server.
+
+To find the time offset between the device and the main server, we send a message with the current device timestamp. The server will then reply with the original timestamp along with the timestamp of the server. We use the response to calculate the actual difference in time.
+
+```js
+var currentTime = Date.now();
+var latency = Math.round((currentTime - e.time) * .5);
+var serverTime = e.serverTime;
+currentTime -= latency;
+var difference = currentTime - serverTime;
+```
+
+Doing this once is not enough, as the round trip to the server is not always symmetrical, meaning it may take longer for the response to get to the server than it does for the server to return it. To get around this, we poll the server multiple times, taking the median result. This gets us within 10ms of the actual difference between device and server.
+
+### Acceleration/Deceleration
+
+When Player 1 presses or releases the screen, the acceleration event is sent to the server. Once received, the server adds its current timestamp and then passes that data along to every other player.
+
+When an “accelerate on” or “accelerate off” event is received by a device, we are able to use the server offset (calculated above) to find out how long it took for that message to be received. This is useful, because Player 1 may receive the message in 20ms, but Player 2 might take 50ms to receive it. This would result in the car being in two different places because device 1 would start the acceleration sooner.
```

Comment:
```suggestion
When an "accelerate on" or "accelerate off" event is received by a device, we are able to use the server offset (calculated above) to find out how long it took for that message to be received. This is useful, because Player 1 may receive the message in 20ms, but Player 2 might take 50ms to receive it. This would result in the car being in two different places because device 1 would start the acceleration sooner.
```